The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6690959150/job/18177258308) [(plain)](https://github.com/rust-lang/rust/commit/a23678c1b89acb052f8d9691007c774be4d5ef61/checks/18177258308/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=roblabla
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_b7c6c9bc-7e58-44c6-a23d-dd3a567fe51b
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=fix-switch-stdio-win7
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_b7c6c9bc-7e58-44c6-a23d-dd3a567fe51b
GITHUB_REF=refs/pull/117386/merge
GITHUB_REF_NAME=117386/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=a617969d20114fb9bb47e2ff5fda4fe89b52aace
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_b7c6c9bc-7e58-44c6-a23d-dd3a567fe51b
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_b7c6c9bc-7e58-44c6-a23d-dd3a567fe51b
GITHUB_TRIGGERING_ACTOR=roblabla
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/117386/merge
GITHUB_WORKFLOW_SHA=a617969d20114fb9bb47e2ff5fda4fe89b52aace
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
   Compiling rand v0.8.5
   Compiling rand_xorshift v0.3.0
   Compiling alloc v0.0.0 (/checkout/library/alloc)
   Compiling std v0.0.0 (/checkout/library/std)
error[E0425]: cannot find value `orig_hdl` in this scope
   |
   |
25 |         assert_ne!(orig_hdl, -1);
   |                    ^^^^^^^^ help: a local variable with a similar name exists: `orig_fd`
For more information about this error, try `rustc --explain E0425`.
For more information about this error, try `rustc --explain E0425`.
error: could not compile `std` (test "switch-stdout") due to previous error
Build completed unsuccessfully in 0:00:39
  local time: Mon Oct 30 09:53:26 UTC 2023
  network time: Mon, 30 Oct 2023 09:53:26 GMT
##[error]Process completed with exit code 1.
```

</details>

--- a/library/std/tests/switch-stdout.rs
+++ b/library/std/tests/switch-stdout.rs
@@ -5,32 +5,48 @@ use std::io::{Read, Write};
 
 mod common;
 
+#[cfg(windows)]
+use std::os::windows::io::OwnedHandle;
+
 #[cfg(unix)]
-fn switch_stdout_to(file: File) {
+use std::os::fd::OwnedFd;
+
+#[cfg(unix)]
+fn switch_stdout_to(file: OwnedFd) -> OwnedFd {
     use std::os::unix::prelude::*;
 
     extern "C" {
+        fn dup(old: i32) -> i32;
         fn dup2(old: i32, new: i32) -> i32;
     }
 
     unsafe {
+        let orig_fd = dup(1);
+        assert_ne!(orig_fd, -1);
+        let res = OwnedFd::from_raw_fd(orig_fd);
         assert_eq!(dup2(file.as_raw_fd(), 1), 1);
+        res
     }
 }
 
 #[cfg(windows)]
-fn switch_stdout_to(file: File) {
+fn switch_stdout_to(file: OwnedHandle) -> OwnedHandle {
     use std::os::windows::prelude::*;
 
     extern "system" {
+        fn GetStdHandle(nStdHandle: u32) -> *mut u8;
         fn SetStdHandle(nStdHandle: u32, handle: *mut u8) -> i32;
     }
 
     const STD_OUTPUT_HANDLE: u32 = (-11i32) as u32;
+    const INVALID_HANDLE_VALUE: *mut u8 = !0 as *mut u8;
 
     unsafe {
+        let orig_hdl = GetStdHandle(STD_OUTPUT_HANDLE);
+        assert!(!orig_hdl.is_null() && orig_hdl != INVALID_HANDLE_VALUE);
         let rc = SetStdHandle(STD_OUTPUT_HANDLE, file.into_raw_handle() as *mut _);
         assert!(rc != 0);
+        OwnedHandle::from_raw_handle(orig_hdl as _)
     }
 }
 
@@ -43,9 +59,11 @@ fn switch_stdout() {
     let mut stdout = std::io::stdout();
     stdout.write(b"foo\n").unwrap();
     stdout.flush().unwrap();
-    switch_stdout_to(f);
+    let orig_hdl = switch_stdout_to(f.into());
     stdout.write(b"bar\n").unwrap();
     stdout.flush().unwrap();
+
+    switch_stdout_to(orig_hdl);
 
     let mut contents = String::new();
     File::open(&path).unwrap().read_to_string(&mut contents).unwrap();

